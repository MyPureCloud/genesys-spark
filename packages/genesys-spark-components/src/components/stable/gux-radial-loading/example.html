<h1>gux-radial-loading</h1>
<div>
  <h2>Default (Modal)</h2>
  <gux-radial-loading screenreader-text="Loading..."></gux-radial-loading>
</div>

<div>
  <h2>Full Page</h2>
  <gux-radial-loading
    context="full-page"
    screenreader-text="Loading..."
  ></gux-radial-loading>
</div>

<div>
  <h2>Input</h2>
  <gux-radial-loading
    context="input"
    screenreader-text="Loading..."
  ></gux-radial-loading>
</div>

<div>
  <h2>Modal</h2>
  <gux-radial-loading
    context="modal"
    screenreader-text="Loading..."
  ></gux-radial-loading>
</div>

<div>
  <h2>Around Icon</h2>
  <div class="around-icon-container">
    <div class="around-icon-container-item">
      <p>Default</p>
      <div class="loading-icon-container">
        <gux-radial-loading
          context="around-icon"
          screenreader-text="Loading..."
        ></gux-radial-loading>
        <gux-icon
          icon-name="fa/phone-regular"
          size="small"
          screenreader-text="Around icon example"
        ></gux-icon>
      </div>
    </div>
    <div class="around-icon-container-item">
      <p>Gux-Button-Slot</p>
      <gux-button-slot icon-only>
        <button type="button" class="loading-icon-container">
          <gux-radial-loading
            context="around-icon"
            screenreader-text="Loading..."
          ></gux-radial-loading>
          <gux-icon
            icon-name="fa/phone-regular"
            size="small"
            screenreader-text="Around icon example"
          ></gux-icon>
        </button>
      </gux-button-slot>
    </div>
  </div>
</div>

<div>
  <h2>Around Icon Light</h2>
  <div class="around-icon-container">
    <div class="around-icon-container-item">
      <p>Default</p>
      <div class="loading-icon-container light dark-bg">
        <gux-radial-loading
          context="around-icon-light"
          screenreader-text="Loading..."
        >
        </gux-radial-loading>
        <gux-icon
          icon-name="fa/phone-regular"
          size="small"
          screenreader-text="Around icon example"
        ></gux-icon>
      </div>
    </div>
    <div class="around-icon-container-item">
      <p>Gux-Button-Slot</p>
      <gux-button-slot icon-only accent="primary">
        <button type="button" class="loading-icon-container light">
          <gux-radial-loading
            context="around-icon-light"
            screenreader-text="Loading..."
          >
          </gux-radial-loading>
          <gux-icon
            icon-name="fa/phone-regular"
            size="small"
            screenreader-text="Around icon example"
          ></gux-icon>
        </button>
      </gux-button-slot>
    </div>
  </div>
</div>

<style>
  .around-icon-container {
    display: flex;
    flex-direction: row;
    gap: 12px;
  }

  .around-icon-container-item {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .loading-icon-container {
    display: grid;
    place-items: center center;
    inline-size: fit-content;
    padding: 0;
  }

  .dark-bg {
    background-color: var(--gse-core-color-mineral-1000);
  }

  .loading-icon-container > * {
    grid-area: 1 / 1;
  }

  .loading-icon-container > gux-icon {
    color: var(--gse-ui-button-secondary-default-foregroundColor);
  }

  .loading-icon-container.light > gux-icon {
    color: var(--gse-ui-button-primary-default-foregroundColor);
  }
</style>
