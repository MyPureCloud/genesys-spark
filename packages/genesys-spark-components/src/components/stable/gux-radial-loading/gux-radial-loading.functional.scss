@use 'sass:math';
@use '~genesys-spark-tokens/dist/scss/gse-ui-feedback';

$largeSpinnerSize: gse-ui-feedback.$gse-ui-progressAndLoading-spinner-large;
$largeSpinnerBorder: gse-ui-feedback.$gse-ui-progressAndLoading-largeBorder;

$smallSpinnerSize: gse-ui-feedback.$gse-ui-progressAndLoading-spinner-small;
$smallSpinnerBorder: gse-ui-feedback.$gse-ui-progressAndLoading-smallBorder;

$thinSpinnerBorder: gse-ui-feedback.$gse-ui-progressAndLoading-smallBorder;

$largeCircumference: (math.$pi * ($largeSpinnerSize - $largeSpinnerBorder));
$smallCircumference: (math.$pi * ($smallSpinnerSize - $smallSpinnerBorder));
$largeThinCircumference: (math.$pi * ($largeSpinnerSize - $thinSpinnerBorder));

@mixin spin($name, $circumference) {
  @keyframes #{$name}Spin {
    0% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(0deg);
    }

    7.5% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(0deg);
    }

    20% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(90deg);
    }

    30% {
      stroke-dasharray: math.div($circumference, 4);
      transform: rotate(180deg);
    }

    40% {
      stroke-dasharray: math.div($circumference, 4);
    }

    50% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(450deg);
    }

    57.5% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(450deg);
    }

    70% {
      stroke-dasharray: math.div($circumference, 4);
      transform: rotate(540deg);
    }

    80% {
      stroke-dasharray: math.div($circumference, 4);
      transform: rotate(630deg);
    }

    90% {
      stroke-dasharray: math.div($circumference, 4);
    }

    100% {
      stroke-dasharray: 1 math.div($circumference, 2) - 1;
      transform: rotate(900deg);
    }
  }
}

@include spin('large', $largeCircumference);
@include spin('small', $smallCircumference);
@include spin('largeThin', $largeThinCircumference);

@mixin gux-spinner-state-style {
  div[role='progressbar'] {
    .gux-svg-container {
      display: block;
      width: calc(var(--gse-ui-progressAndLoading-spinner-large) * 1px);
      height: calc(var(--gse-ui-progressAndLoading-spinner-large) * 1px);

      .gux-dynamic-circle,
      .gux-dynamic-circle-mask {
        fill: none;
        stroke: var(--gse-ui-progressAndLoading-spinner-foreground);
        stroke-width: $largeSpinnerBorder;
        transform-origin: 50% 50%;
        animation-name: largeSpin;
        animation-duration: 2s;
        animation-timing-function: linear;
        animation-iteration-count: infinite, infinite;
      }

      .gux-dynamic-circle-mask {
        display: none;
        stroke: white;
        stroke-width: $largeSpinnerBorder - $thinSpinnerBorder;
      }

      .gux-static-circle {
        fill: none;
        stroke: var(--gse-ui-progressAndLoading-spinner-base);
        stroke-width: $largeSpinnerBorder;
      }
    }

    &.gux-full-page {
      .gux-svg-container {
        .gux-dynamic-circle-mask {
          display: initial;
        }

        .gux-static-circle {
          display: none;
        }
      }
    }

    &.gux-input {
      .gux-svg-container {
        width: calc(var(--gse-ui-progressAndLoading-spinner-small) * 1px);
        height: calc(var(--gse-ui-progressAndLoading-spinner-small) * 1px);

        .gux-dynamic-circle {
          fill: none;
          stroke: var(--gse-ui-progressAndLoading-spinner-foreground);
          stroke-width: $smallSpinnerBorder;
          transform-origin: 50% 50%;
          animation-name: smallSpin;
          animation-duration: 2s;
          animation-timing-function: linear;
          animation-iteration-count: infinite, infinite;
        }

        .gux-static-circle {
          stroke: var(--gse-ui-progressAndLoading-spinner-base);
          stroke-width: $smallSpinnerBorder;
        }
      }
    }
  }
}
