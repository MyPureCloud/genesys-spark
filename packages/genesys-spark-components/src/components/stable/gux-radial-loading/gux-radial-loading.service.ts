export const largeSpinner = 48; // Linked to --gse-ui-progressAndLoading-spinner-large
export const largeBorder = 4; // Linked to --gse-ui-progressAndLoading-largeBorder

export const mediumSpinner = 30; // Linked to --gse-ui-progressAndLoading-spinner-medium

export const smallSpinner = 16; // Linked to --gse-ui-progressAndLoading-spinner-small
export const smallBorder = 2; // Linked to --gse-ui-progressAndLoading-smallBorder
