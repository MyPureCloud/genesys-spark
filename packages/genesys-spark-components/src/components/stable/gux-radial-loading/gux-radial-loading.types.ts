export type GuxRadialLoadingContext =
  | 'full-page'
  | 'input'
  | 'modal'
  | 'around-icon'
  | 'around-icon-light';
