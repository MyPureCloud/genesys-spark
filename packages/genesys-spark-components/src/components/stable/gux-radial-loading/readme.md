# gux-radial-loading

When the loading of a task is known, this component displays a radial indicator and the completed percentage of
the task.

Otherwise, if the loading is unknown, and no percentage is provided, it instead displays a simple Loading Spinner.

<!-- Auto Generated Below -->


## Properties

| Property           | Attribute           | Description                                                                                                                                                   | Type                                | Default   |
| ------------------ | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------- | --------- |
| `context`          | `context`           | The display context the component is in.                                                                                                                      | `"full-page" \| "input" \| "modal"` | `'modal'` |
| `screenreaderText` | `screenreader-text` | Localized text to provide an accessible label for the component. If no screenreader text is provided, the localized string "Loading" will be used by default. | `string`                            | `''`      |


## Dependencies

### Used by

 - [gux-dropdown](../gux-dropdown)
 - [gux-dropdown-multi](../gux-dropdown-multi)
 - [gux-listbox](../gux-listbox)
 - [gux-listbox-multi](../gux-listbox-multi)
 - [gux-page-loading-spinner](../gux-page-loading-spinner)
 - [gux-radial-progress](../gux-radial-progress)
 - [gux-toggle](../gux-toggle)

### Graph
```mermaid
graph TD;
  gux-dropdown --> gux-radial-loading
  gux-dropdown-multi --> gux-radial-loading
  gux-listbox --> gux-radial-loading
  gux-listbox-multi --> gux-radial-loading
  gux-page-loading-spinner --> gux-radial-loading
  gux-radial-progress --> gux-radial-loading
  gux-toggle --> gux-radial-loading
  style gux-radial-loading fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
