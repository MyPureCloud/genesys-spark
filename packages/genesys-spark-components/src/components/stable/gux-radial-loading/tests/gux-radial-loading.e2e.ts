import { E2EElement } from '@stencil/core/testing';
import { newSparkE2EPage, a11yCheck } from '../../../../test/e2eTestUtils';
import { renderConfigs } from './gux-radial-loading.common';

function getInternalProgressBar(radialLoadingElement: E2EElement): Element {
  return radialLoadingElement.shadowRoot.querySelector(
    'div[role="progressbar"]'
  );
}

describe('gux-radial-loading', () => {
  renderConfigs.forEach(({ description, html }) => {
    it(description, async () => {
      const page = await newSparkE2EPage({ html });
      const element = await page.find('gux-radial-loading');

      await a11yCheck(page);
      expect(element.outerHTML).toMatchSnapshot();
    });
  });

  it('should add an aria-label with the provided screenreader-text to the progressbar', async () => {
    const html =
      '<gux-radial-loading lang="en" screenreader-text="Loading Content"></gux-radial-loading>';
    const page = await newSparkE2EPage({ html });
    const element = await page.find('gux-radial-loading');
    await a11yCheck(page);

    expect(getInternalProgressBar(element).getAttribute('aria-label')).toEqual(
      'Loading Content'
    );
  });

  it('should add the default aria-label text if no screenreader-text is provided', async () => {
    const html = '<gux-radial-loading lang="en"></gux-radial-loading>';
    const page = await newSparkE2EPage({ html });
    const element = await page.find('gux-radial-loading');
    await a11yCheck(page);

    expect(getInternalProgressBar(element).getAttribute('aria-label')).toEqual(
      'Loading'
    );
  });
});
