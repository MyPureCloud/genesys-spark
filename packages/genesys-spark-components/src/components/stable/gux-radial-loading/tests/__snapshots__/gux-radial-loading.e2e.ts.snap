// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-radial-loading should render as expected for "default" context 1`] = `"<gux-radial-loading lang="en" context="default" hydrated=""><template shadowrootmode="open"><div class="gux-default" role="progressbar" aria-label="Loading"><svg class="gux-svg-container" viewBox="0 0 48 48" role="presentation"><circle cx="50%" cy="50%" r="22" class="gux-static-circle"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle" stroke-linecap="round"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle-mask" stroke-linecap="round"></circle></svg></div></template></gux-radial-loading>"`;

exports[`gux-radial-loading should render as expected for "full-page" context 1`] = `"<gux-radial-loading lang="en" context="full-page" hydrated=""><template shadowrootmode="open"><div class="gux-full-page" role="progressbar" aria-label="Loading"><svg class="gux-svg-container" viewBox="0 0 48 48" role="presentation"><circle cx="50%" cy="50%" r="22" class="gux-static-circle"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle" stroke-linecap="round"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle-mask" stroke-linecap="round"></circle></svg></div></template></gux-radial-loading>"`;

exports[`gux-radial-loading should render as expected for "input" context 1`] = `"<gux-radial-loading lang="en" context="input" hydrated=""><template shadowrootmode="open"><div class="gux-input" role="progressbar" aria-label="Loading"><svg class="gux-svg-container" viewBox="0 0 16 16" role="presentation"><circle cx="50%" cy="50%" r="7" class="gux-static-circle"></circle><circle cx="50%" cy="50%" r="7" class="gux-dynamic-circle" stroke-linecap="round"></circle><circle cx="50%" cy="50%" r="7" class="gux-dynamic-circle-mask" stroke-linecap="round"></circle></svg></div></template></gux-radial-loading>"`;

exports[`gux-radial-loading should render as expected for "modal" context 1`] = `"<gux-radial-loading lang="en" context="modal" hydrated=""><template shadowrootmode="open"><div class="gux-modal" role="progressbar" aria-label="Loading"><svg class="gux-svg-container" viewBox="0 0 48 48" role="presentation"><circle cx="50%" cy="50%" r="22" class="gux-static-circle"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle" stroke-linecap="round"></circle><circle cx="50%" cy="50%" r="22" class="gux-dynamic-circle-mask" stroke-linecap="round"></circle></svg></div></template></gux-radial-loading>"`;
